### ISSUE TITLE
Problem with using conditional mux execution with nans

### ISSUE THREAD

**Issue by KsawerySmoczynski** [NVIDIA/DALI#3954]
I have such problem that via external operator I'm decoding 1-d images that contain nan values within them. For certain operations such as slice, rotate etc. I can specify the fill value which however cannot be np.nan. After running the pipeline I'd like retrieve image with all original nans and ones created with augmentations. For this purpose I'm using some specific defined value for fill, so later I will be able to map it to nans.
I was trying to recover nan values by using [mux as described here](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/expressions/expr_conditional_and_masking.html#Multiplexing-as-a-Helper-Function), using boolean mask obtained by comparing image with value that I predefined for fill.
The problem is that when I use constant filled with nans it turns anything into nans after performing math operation against it, resulting in whole image filled with nans.

Pipeline returning nans only. Not using mux block results in imgs returned with nans where they were and no_data_value for filled pixels:
```python
@pipeline_def
def EnhancedSegmentationPipeline(
    data_iterator: Callable,
    img_size: int,
    no_data_value: int,
):
   img = fn.external_source(source=data_iterator, batch=False, \
                    parallel=True, dtype=dali.types.FLOAT)
   rotation_angle = fn.random.uniform(range=(-20, 20))
   img = fn.rotate(
            img,
            angle=rotation_angle,
            fill_value=no_data_value,
            interp_type=dali.types.INTERP_LINEAR,
            dtype=dali.types.FLOAT,
        ) 
   img_nan_mask = fn.constant(fdata=no_data_value, shape=(img_size, img_size, 1),  dtype=dali.types.FLOAT)
   img_nan_mask = img == img_nan_mask
   img = mux(
        img_nan_mask,
        fn.constant(fdata=np.nan, shape=(img_size, img_size, 1), dtype=dali.types.FLOAT),
        img
    )
   return img
```
I know that I could handle it outside of the pipeline, but I would rather keep it encapsulated within. Is it possible?

**Comment by mzient** [comment]
The `mux` thingy is not something to be used unless you really, really have to. It's a workaround and a hack.
Having said that, you can make it even more hacky and terrible, but get it to work by multiplexing _integers_. All you need to do is reinterpret (`fn.reinterpret`) your float data as `dali.types.INT32`, multiplex that and then reinterpret back.

**Comment by KsawerySmoczynski** [comment]
@mzient thanks for your reply!

I tried to do it like this according to your suggestion:

```python
@pipeline_def
def EnhancedPipeline(
    data_iterator: Callable,
    img_size: int,
    no_data_value: int,
):
   img = fn.external_source(source=data_iterator, batch=False, \
                    parallel=True, dtype=dali.types.FLOAT)
   rotation_angle = fn.random.uniform(range=(-20, 20))
   img = fn.rotate(
            img,
            angle=rotation_angle,
            fill_value=no_data_value,
            interp_type=dali.types.INTERP_LINEAR,
            dtype=dali.types.FLOAT,
        ) 

   img_nan_mask = fn.constant(idata=no_data_value, shape=(img_size, img_size, 1), device=device, dtype=dali.types.UINT32)
   img = fn.reinterpret(dsms, shape=(img_size, img_size, 1), dtype=dali.types.UINT32)
   img_nan_mask = img == img_nan_mask
   img = mux(
        img_nan_mask,
        fn.reinterpret(
            fn.constant(fdata=np.nan, shape=(img_size, img_size, 1), device=device, dtype=dali.types.FLOAT), 
            shape=(img_size, img_size, 1), 
            dtype=dali.types.UINT32
        ),
        img
    )
   img = fn.reinterpret(img, shape=(img_size, img_size, 1), dtype=dali.types.FLOAT)

   return img
```

But it doesn't map anything to np.nan, the no_data_value persists after this operation, any ideas why it might be the case?

offtop (my use case for mux):
When it comes to mux, I see no other mechanism of mapping values currently, especially for floats (as fn.lookup table works for integer keys only and all range of values have to be provided upfront, you cannot map a single value). For example grid mask augmentation fills image with 0s, so in case of semantic segmentation pipeline the applying it jointly to images and labels results in latter ones filled with zero, which can be arbitrary class. In this case I use mux to remap it to "background" or "no_data" class indicating loss masking.

**Comment by mzient** [comment]
1. What is `dsms` in this line: `   img = fn.reinterpret(dsms, shape=(img_size, img_size, 1), dtype=dali.types.UINT32)` ?
2. As per the original code - you can't detect nans by comparing something to NaN. Two floats with NaN values don't yield true in _any_ comparison. Actually, to detect NaNs in the image you should do:
```
not_nan = img == img
```
The NaN entries will not compare equal to themselves.

I've modified your original code:

```Python
@pipeline_def
def EnhancedSegmentationPipeline(
    data_iterator: Callable,
    img_size: int,
    no_data_value: int,
):
   img = fn.external_source(source=data_iterator, batch=False, \
                    parallel=True, dtype=dali.types.FLOAT)
   rotation_angle = fn.random.uniform(range=(-20, 20))
   img = fn.rotate(
            img,
            angle=rotation_angle,
            fill_value=no_data_value,
            interp_type=dali.types.INTERP_LINEAR,
            dtype=dali.types.FLOAT,
        ) 
   img_not_nan_mask = img == img  # detect not-nan values will be True
   img = mux(
        img_not_nan_mask,
        fn.reinterpret(img, dtype=dali.types.UINT32),
        fn.reinterpret(fn.constant(fdata=np.nan, shape=(img_size, img_size, 1), dtype=dali.types.FLOAT), dtype=dali.types.UINT32)
    )
   img = fn.reinterpret(img, dtype=dali.types.FLOAT)
   return img
```

**Comment by KsawerySmoczynski** [comment]
1. Should be img, updated my comment.

2. I'm trying to detect the value introduced by augmentation with fill_value argument and then replace it with nans. I'm not touching the original nans loaded with the image. The name of the variable was misleading I guess, it should be probably no_data_value_mask.

```python
@pipeline_def
def EnhancedPipeline(
    data_iterator: Callable,
    img_size: int,
    no_data_value: int,
):
   img = fn.external_source(source=data_iterator, batch=False, \
                    parallel=True, dtype=dali.types.FLOAT)
   rotation_angle = fn.random.uniform(range=(-20, 20))
   img = fn.rotate(
            img,
            angle=rotation_angle,
            fill_value=no_data_value,
            interp_type=dali.types.INTERP_LINEAR,
            dtype=dali.types.FLOAT,
        ) 

   no_data_value_mask = fn.constant(idata=no_data_value, shape=(img_size, img_size, 1), device=device, dtype=dali.types.UINT32)
   img = fn.reinterpret(img, shape=(img_size, img_size, 1), dtype=dali.types.UINT32)
   no_data_value_mask = img == no_data_value_mask
   img = mux(
        no_data_value_mask,
        fn.reinterpret(
            fn.constant(fdata=np.nan, shape=(img_size, img_size, 1), device=device, dtype=dali.types.FLOAT), 
            shape=(img_size, img_size, 1), 
            dtype=dali.types.UINT32
        ),
        img
    )
   img = fn.reinterpret(img, shape=(img_size, img_size, 1), dtype=dali.types.FLOAT)

   return img
```

**Comment by JanuszL** [comment]
@KsawerySmoczynski,

Can you make this example self-contained so we can run this on our own (you can use https://github.com/NVIDIA/DALI_extra as a data source or provide a test file that would illustrate the problem)?

**Comment by KsawerySmoczynski** [comment]
Below is the example with applied suggestion regarding reinterpreting img as a UINT32. 
Applying mapping can be altered with APPLY_MAPPING constant. 

```python
import random
from pathlib import Path
from typing import Callable, List
import numpy as np
from PIL import Image
import cv2

from nvidia import dali
from nvidia.dali import pipeline_def, Pipeline
from nvidia.dali import fn
from nvidia.dali.plugin.pytorch import DALIGenericIterator
from nvidia.dali import pickling as dali_pickle
import torch


@dali_pickle.pickle_by_value
def SegmentationCallback(
    imgs_list: List[str], batch_size: int, seed: int, shard_id: int = 0, num_shards: int = 1, **kwargs
):
    indices = np.arange(len(imgs_list))
    np.random.shuffle(indices)

    dataset_size = len(imgs_list)
    shard_size = dataset_size // num_shards
    shard_offset = shard_size * shard_id
    n_iterations = shard_size // batch_size
    perm = None
    last_seen_epoch = None

    def callback(sample_info):
        nonlocal perm, last_seen_epoch
        if sample_info.iteration >= n_iterations:
            raise StopIteration
        if last_seen_epoch != sample_info.epoch_idx:
            last_seen_epoch = sample_info.epoch_idx
            perm = np.random.default_rng(seed=seed + sample_info.epoch_idx).permutation(dataset_size)

        index = perm[sample_info.idx_in_epoch + shard_offset]
        img = cv2.imread(imgs_list[index], cv2.IMREAD_ANYDEPTH)[..., np.newaxis]
        return img

    return callback


def mux(condition, true_case, false_case):
    neg_condition = condition ^ True
    return condition * true_case + neg_condition * false_case


@pipeline_def
def EnhancedPipeline(data_iterator: Callable, img_size: int, no_data_value: int, apply_mapping: bool):

    use_gpu = Pipeline.current().device_id is not None

    img = fn.external_source(source=data_iterator, batch=False, parallel=True, dtype=dali.types.FLOAT)
    if use_gpu:
        img.gpu()

    rotation_angle = fn.random.uniform(range=(-20, 20))
    img = fn.rotate(
        img,
        angle=rotation_angle,
        fill_value=no_data_value,
        interp_type=dali.types.INTERP_LINEAR,
        dtype=dali.types.FLOAT,
    )

    size = np.random.uniform(int(img_size * 0.75), int(img_size * 1.25))
    border = int(img_size * 0.25)
    crop_pos_x = np.random.uniform(-border, img_size - (size - border))
    crop_pos_y = np.random.uniform(-border, img_size - (size - border))

    img = fn.slice(
        img,
        start=[crop_pos_x, crop_pos_y],
        shape=[size, size],
        axis_names="HW",
        out_of_bounds_policy="pad",
        fill_values=no_data_value,
    )

    img = fn.resize(
        img,
        size=(img_size, img_size),
        interp_type=dali.types.INTERP_LINEAR,
    )

    if apply_mapping:
        no_data_value_mask = fn.reinterpret(
            fn.constant(fdata=no_data_value, shape=(img_size, img_size, 1), dtype=dali.types.FLOAT),
            shape=(img_size, img_size, 1),
            dtype=dali.types.UINT32,
        )
        img = fn.reinterpret(img, shape=(img_size, img_size, 1), dtype=dali.types.UINT32)
        no_data_value_mask = img == no_data_value_mask  # True for no_data_value introduced by augmentation
        img = mux(
            no_data_value_mask,
            fn.reinterpret(
                fn.constant(fdata=np.nan, shape=(img_size, img_size, 1), dtype=dali.types.FLOAT),
                shape=(img_size, img_size, 1),
                dtype=dali.types.UINT32,
            ),
            img,
        )
        img = fn.reinterpret(img, shape=(img_size, img_size, 1), dtype=dali.types.FLOAT)

    return img


if __name__ == "__main__":
    seed = 42
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

    APPLY_MAPPING = True
    NO_DATA_VALUE = -10000.

    img = np.zeros((512, 512)) + np.nan
    img[256:, 256:] = 100
    img[384:, 384:] = 200

    data_root = Path("imgs")
    data_root.mkdir(parents=True, exist_ok=True)

    Image.fromarray(img, mode="F").save(data_root / "1.tif", compression="tiff_lzw")

    imgs_list = [str(data_root / "1.tif")] * 10

    batch_size = 1
    num_workers = 2
    pipeline_config = {
        "batch_size": batch_size,
        "device_id": None if not torch.cuda.is_available() else 0,
        "seed": seed,
        "num_threads": num_workers,
        "py_num_workers": num_workers,
        "py_start_method": "spawn",
    }

    pipe = EnhancedPipeline(
        SegmentationCallback(imgs_list, batch_size, seed), 512, NO_DATA_VALUE, APPLY_MAPPING, **pipeline_config
    )
    pipe.build()

    iterator = DALIGenericIterator(pipe, output_map=["img"])

    batch = next(iterator)

    img = batch[0]["img"][0]

    print(f"Nans ratio {'with' if APPLY_MAPPING else 'without'} mapping {img.isnan().sum()/len(img.flatten())}")
    print(f"#NO_DATA_VALUE {'with' if APPLY_MAPPING else 'without'} mapping {(img == NO_DATA_VALUE).sum()}")
```

With no_data_value_mask is defined as:
```python
no_data_value_mask = fn.reinterpret(
            fn.constant(fdata=float(no_data_value), shape=(img_size, img_size, 1), dtype=dali.types.FLOAT),
            shape=(img_size, img_size, 1),
            dtype=dali.types.UINT32,
        )
``` 
the behaviour of mapping everything to nans persists.

With no_data_value_mask defined as:
```python
no_data_value_mask = fn.constant(idata=int(no_data_value), shape=(img_size, img_size, 1), dtype=dali.types.UINT32))
```
nothing gets mapped as comparison yields false everywhere.

**Comment by KsawerySmoczynski** [comment]
Great, it's working perfectly, thanks a lot for your help!
